server/src/main/java/org/opensearch/index/translog/transfer/TranslogTransferManager.java

```diff
@@ -101,44 +111,54 @@ public ShardId getShardId() {
-            if (toUpload.isEmpty()) {
+            int totalFilesCount = transferSnapshot.getTranslogTransferMetadata().getCount();
+            List<Exception> exceptionList = new ArrayList<>(totalFilesCount);
+            Set<TranslogCheckpointSnapshot> generationalSnapshotList = new HashSet<>(totalFilesCount);
+            generationalSnapshotList.addAll(fileTransferTracker.exclusionFilter(transferSnapshot.getTranslogCheckpointSnapshots()));
+
+            if (generationalSnapshotList.isEmpty()) {
                 logger.trace("Nothing to upload for transfer");
                 return true;
             }
 
-            fileTransferTracker.recordBytesForFiles(toUpload);
+            fileTransferTracker.recordBytesForFiles(generationalSnapshotList);
             captureStatsBeforeUpload();
-            final CountDownLatch latch = new CountDownLatch(toUpload.size());
-            LatchedActionListener<TransferFileSnapshot> latchedActionListener = new LatchedActionListener<>(
+            final CountDownLatch latch = new CountDownLatch(generationalSnapshotList.size());
+            LatchedActionListener<TranslogCheckpointSnapshot> latchedActionListener = new LatchedActionListener<>(
                 ActionListener.wrap(fileTransferTracker::onSuccess, ex -> {
-                    assert ex instanceof FileTransferException;
+                    assert ex instanceof TranslogGenerationTransferException;
                     logger.error(
                         () -> new ParameterizedMessage(
-                            "Exception during transfer for file {}",
-                            ((FileTransferException) ex).getFileSnapshot().getName()
+                            "Exception during transfer for translog generation {}",
+                            ((TranslogGenerationTransferException) ex).getFileSnapshot().getGeneration()
                         ),
                         ex
                     );
-                    FileTransferException e = (FileTransferException) ex;
-                    TransferFileSnapshot file = e.getFileSnapshot();
+                    TranslogGenerationTransferException e = (TranslogGenerationTransferException) ex;
+                    TranslogCheckpointSnapshot file = e.getFileSnapshot();
                     fileTransferTracker.onFailure(file, ex);
                     exceptionList.add(ex);
+
+                    Set<TransferFileSnapshot> failedFiles = e.getFailedFiles();
+                    Set<TransferFileSnapshot> successFiles = e.getSuccessFiles();
+                    if (!failedFiles.isEmpty()) {
+                        failedFiles.forEach(failedFile -> { fileTransferTracker.add(failedFile.getName(), false); });
+                    }
+                    if (!successFiles.isEmpty()) {
+                        successFiles.forEach(successFile -> { fileTransferTracker.add(successFile.getName(), true); });
+                    }
```

Comment:
You can have an abstract method that returns a boolean and use that to decide if the file name based tracker should run or not.

Reply:
Also, failedFiles has to be non-empty here..

Reply:
If you return non-null, you dont need the `if` check.